src/main/java/com/google/sps/data/DatastorePersonDao.java

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Accesses Datastore to support managing Person entities. */
+public class DatastorePersonDao implements PersonDao {
+  // @param datastore the DatastoreService we're using to interact with Datastore.
+  private DatastoreService datastore;
+
+  /** Initializes the fields for PersonDatastoreDAO. */
+  public DatastorePersonDao() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /* We make an entity in Datastore with person's fields as properties and their email as the key. */
+  public void put(Person person) {
+    datastore.put(personToEntity(person));
+  }
+
+  private static Entity personToEntity(Person person) {
+    Entity personEntity = new Entity("Person", person.key());
+    personEntity.setProperty("email", person.key());    
+    personEntity.setProperty("firstName", person.getFirstName());
+    personEntity.setProperty("lastName", person.getLastName());
+    personEntity.setProperty("company", person.getCompany());
+    personEntity.setProperty("job", person.getJob());
+    personEntity.setProperty("linkedIn", person.getLinkedIn());
+    return personEntity;
+  }
+
+  /* Since email is a personEntity's key, we retrieve the corresponding personEntity from Datastore and return it as a Person object. */
+  // TODO: make our own PersonNotFoundException and refactor how we're handling these
+  public Person get(String email)
+      throws com.google.appengine.api.datastore.EntityNotFoundException {
```

Comment:
One option here is to return Optional<Person> from get.

To do that, catch EntityNotFoundException in the implementation and return Optional.empty().

(the type is java.util.Optional).

Reply:
lf l return an Optional, then higher up in the call chain (in the PersonServlet) l have to check if the Optional is empty and then do my business logic (redirect to registration page). ls this what you want?

Reply:
Having a dedicated exception for handling "personnotfound" doesn't buy you much. I think it's really only useful in this one context.

It could make sense to have an exception here, if you never expect the person to be missing. That said, I think Optional better expresses the semantics. If the caller cannot handle the absence of the value (i.e. it cannot render a servlet without it), then the Optional is exactly what you want, because it forces you to handle this case directly instead of just pushing the exception upward.